Support RGB, YBR_FULL and YBR_FULL_422 PhotoMetricInterpretation

pevers opened the pull request:
Hi @Enet4 

This PR adds support for `RGB`, `YBR_FULL` and `YBR_FULL_422` as PhotoMetricInterpretation.

It works by converting `YBR_FULL` and `YBR_FULL_422` color spaces to RGB.

Now that I read through this PR, I think it might be a better option to do this:

1. Read as `ndarray`.
2. Optional Apply modality LUT and make it a public method under the `utilities` module.
3. Optional Convert color space and make it a public method under the `utilities` module.
4. Convert to `DynamicImage`

What do you think? 

Closes: https://github.com/Enet4/dicom-rs/issues/119

Enet4 reviewed (changes requested):
Thank you for working on this! Things look good so far. I have provided a quick review for now.

> 1. Read as ndarray.
> 2. Optional Apply modality LUT and make it a public method under the utilities module.
> 3. Optional Convert color space and make it a public method under the utilities module.
> 4. Convert to DynamicImage

One could envision a nice API that would chain the operations together like this:

```rust
let arr = obj.pixel_data()
    .with_modality_lut(options)?
    .to_ndarray::<T>()?;
```

```rust
let img = obj.pixel_data()
    .with_modality_voi_lut(options)?
    .to_dynamic_image()?
```

This API would be possible through a pattern similar to that of the consuming iterator. That is, each method would move `self` into something representing the same pixel data with some transformations.

On the other hand, there may be reasonable defaults so that I am not quite sure whether each step would have a different interface at all. One could stop the user from trying to create an image without performing the VOI LUT, but there may be a set of window level attributes in the object that are expected to be applied.

I say that this API can be researched here or in another PR.

pevers commented on pixeldata/src/lib.rs:
```diff
@@ -258,9 +272,46 @@ fn normalize_to_u16(i: &[f64]) -> Vec<u16> {
+}
+
+// Convert a normalized u16 pixel array from YBR_FULL or YBR_FULL_422 to RGB
+fn convert_ybr_full_to_rgb(i: &mut [u16]) -> Vec<u16> {
```
You are right! This would be much more efficient changing the pixel array inline.

pevers commented on pixeldata/src/lib.rs:
```diff
@@ -258,9 +272,46 @@ fn normalize_to_u16(i: &[f64]) -> Vec<u16> {
+            let cr: u16 = cr.floor().clamp(0.0, u16::MAX as f32) as u16;
+            let cg: u16 = cg.floor().clamp(0.0, u16::MAX as f32) as u16;
+            let cb: u16 = cb.floor().clamp(0.0, u16::MAX as f32) as u16;
+            vec![cr, cg, cb]
```
Haha, yep. I'm now changing this to a slightly different approach and also inline.

pevers commented:
Hi @Enet4 , I changed it a bit. 

There were many places where I converted vectors back and forth. I tried to reduce that.

I reduced the complexity a bit by assuming that colored dicoms are either 8 bit, or 16 bit unsigned without any transformation. I think that is a safe assumption.

Grayscale images are handled via `to_ndarray`. Modality LUT is applied, MONOCHROME1 is inverted (I couldn't find a valid test file yet, I'll search for one), and then normalized to u16. VOI Lut is still missing.

I'm also replacing pixels when converting them to RGB, instead of creating a new vector.

Enet4 reviewed (approved):
All right, I noticed that this also implements MONOCHROME1! Things seem to be in order. I will just take the liberty of adding a reference to the issue alongside the TODO comment, as it helps to prevent them from becoming forgotten or inaccurate.

Thank you very much for your continued contribution!
